slidechain/postpegout.go

```diff
@@ -0,0 +1,111 @@
+	if err != nil {
+		return errors.Wrap(err, "unmarshaling asset xdr")
+	}
+	assetID := bc.NewHash(txvm.AssetID(importIssuanceSeed[:], assetXDR))
+	ref := pegOut{
+		AssetXDR: assetXDR,
+		TempAddr: tempAddr,
+		Seqnum:   seqnum,
+		Exporter: exporter,
+		Amount:   amount,
+		Anchor:   anchor,
+		Pubkey:   pubkey,
+	}
+	refdata, err := json.Marshal(ref)
+	if err != nil {
+		return errors.Wrap(err, "marshaling reference data")
+	}
+	// The contract needs a non-zero selector to retire funds if the peg-out succeeded.
+	// Else, it requires a zero selector so the funds are returned.
+	var selector int64
+	if peggedOut == pegOutOK {
+		selector = 1
+	}
+	// Build post-peg-out contract.
+	b := new(txvmutil.Builder)
+	b.Tuple(func(contract *txvmutil.TupleBuilder) { // {'C', ...}
+		contract.PushdataByte(txvm.ContractCode)
+		contract.PushdataBytes(exportContract1Seed[:])
+		contract.PushdataBytes(exportContract2Prog)
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'T', pubkey}
+			tup.PushdataByte(txvm.TupleCode)
+			tup.Tuple(func(pktup *txvmutil.TupleBuilder) {
+				pktup.PushdataBytes(pubkey)
+			})
+		})
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'V', amount, assetID, anchor}
+			tup.PushdataByte(txvm.ValueCode)
+			tup.PushdataInt64(amount)
+			tup.PushdataBytes(assetID.Bytes())
+			tup.PushdataBytes(anchor)
+		})
+		contract.Tuple(func(tup *txvmutil.TupleBuilder) { // {'S', refdata}
+			tup.PushdataByte(txvm.BytesCode)
+			tup.PushdataBytes(refdata)
+		})
+	})
+	b.PushdataInt64(selector).Op(op.Put)                                // con stack: snapshot; arg stack: selector
+	b.Op(op.Input).Op(op.Call)                                          // arg stack: sigchecker
+	b.PushdataBytes(c.InitBlockHash.Bytes())                            // con stack: blockid; arg stack: sigchecker
+	b.PushdataInt64(int64(bc.Millis(time.Now().Add(10 * time.Minute)))) // con stack: blockid, expmss; arg stack: sigchecker
+	b.Op(op.Nonce).Op(op.Finalize)                                      // arg stack: sigchecker
+
+	// Check signature.
+	prog1 := b.Build()
+	vm, err := txvm.Validate(prog1, 3, math.MaxInt64, txvm.StopAfterFinalize)
+	if err != nil {
+		return errors.Wrap(err, "computing transaction ID")
+	}
+	sig := ed25519.Sign(c.privkey, vm.TxID[:])
+	b.Op(op.Get).PushdataBytes(sig).Op(op.Put) // con stack: sigchecker; arg stack: sig
+	b.Op(op.Call)
+
+	// Build, submit, and wait for the post-peg-out tx to hit txvm.
+	prog2 := b.Build()
+	tx, err := bc.NewTx(prog2, 3, math.MaxInt64)
+	if err != nil {
+		return errors.Wrap(err, "making post-peg-out tx")
+	}
+	r, err := c.S.submitTx(ctx, tx)
+	if err != nil {
+		return errors.Wrap(err, "submitting post-peg-out tx")
+	}
+	err = c.S.waitOnTx(ctx, tx.ID, r)
+	if err != nil {
+		return errors.Wrap(err, "waiting on post-peg-out tx to hit txvm")
+	}
+	// Delete relevant row from exports table.
+	result, err := c.DB.ExecContext(ctx, `DELETE FROM exports WHERE txid=$1`, txid)
```

Comment:
What if a crash occurs right before this line? We'll have retired or refunded the txvm funds, but not updated the db. How should slidechaind recover the next time it runs?